File: frontend/src/types/evaluation.ts
// Vue 概念：类型定义让评测配置、请求和指标结果在组件中保持强类型。
// Vue 概念：配置数组由 v-for 渲染，用户可以增删不同检索方案。
// Vue 概念：结果表与样本明细复用同一份响应式数据。

export type SearchMode = 'vector' | 'bm25' | 'hybrid'
export type EvaluationRunStatus = 'queued' | 'running' | 'completed' | 'failed'

export interface EvalDatasetSummary {
  dataset_name: string
  case_count: number
}

export interface RetrievalConfigInput {
  name: string
  mode: SearchMode
  top_k: number
  rerank: boolean
}

export interface EvalRunRequest {
  dataset_name: string
  configs: RetrievalConfigInput[]
}

export interface CaseMetrics {
  question: string
  recall_at_5: number
  reciprocal_rank: number
  faithfulness: number
  answer_relevance: number
  ragas_faithfulness: number | null
  ragas_answer_relevance: number | null
  latency_ms: number
  first_token_latency_ms: number
  input_tokens: number
  output_tokens: number
  estimated_cost: number
  refused: boolean
  error: string | null
}

export interface ConfigMetrics extends RetrievalConfigInput {
  case_count: number
  recall_at_5: number
  mrr: number
  faithfulness: number
  answer_relevance: number
  ragas_faithfulness: number | null
  ragas_answer_relevance: number | null
  average_latency_ms: number
  average_first_token_latency_ms: number
  average_input_tokens: number
  average_output_tokens: number
  average_estimated_cost: number
  refusal_rate: number
  hallucination_risk: number
  cases: CaseMetrics[]
}

export interface EvalRunResponse {
  run_id: string
  dataset_name: string
  task_id: string | null
  status: EvaluationRunStatus
  progress_completed: number
  progress_total: number
  error_message: string | null
  created_at: string
  started_at: string | null
  completed_at: string | null
  results: ConfigMetrics[]
}

export interface EvaluationRunAccepted {
  run_id: string
  dataset_name: string
  task_id: string
  status: EvaluationRunStatus
  progress_completed: number
  progress_total: number
}

export interface EvaluationRunSummary {
  run_id: string
  dataset_name: string
  status: EvaluationRunStatus
  progress_completed: number
  progress_total: number
  created_at: string
}

File: backend/app/schemas/evaluation.py
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.retrieval.types import SearchMode


class EvalDatasetSummary(BaseModel):
    dataset_name: str
    case_count: int


class EvalDatasetVersionResponse(BaseModel):
    id: UUID
    dataset_name: str
    revision: int
    content_hash: str
    case_count: int
    created_by: str
    created_at: str


class EvalCaseResponse(BaseModel):
    id: UUID
    question: str
    reference_answer: str
    expected_chunk_ids: list[UUID]
    tags: list[str]
    created_by: str


class EvalCaseCreateRequest(BaseModel):
    question: str = Field(min_length=1)
    reference_answer: str = Field(min_length=1)
    expected_chunk_ids: list[UUID] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    created_by: str = Field(min_length=1, max_length=255)


class RetrievalConfigRequest(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    mode: SearchMode
    top_k: int = Field(default=5, ge=1, le=50)
    rerank: bool = False


class EvalRunRequest(BaseModel):
    dataset_name: str = Field(min_length=1)
    configs: list[RetrievalConfigRequest] = Field(min_length=1, max_length=8)


class RefusalThresholdCalibrationRequest(BaseModel):
    dataset_name: str = Field(min_length=1)
    config: RetrievalConfigRequest


class RefusalThresholdResponse(BaseModel):
    mode: SearchMode
    top_k: int
    rerank_provider: str
    threshold: float
    sample_count: int
    metrics: dict[str, float | int]
    created_at: str


EvaluationRunStatus = Literal[
    "queued",
    "running",
    "completed",
    "failed",
    "cancelled",
]


class CaseMetricsResponse(BaseModel):
    question: str
    recall_at_5: float
    reciprocal_rank: float
    faithfulness: float
    answer_relevance: float
    ragas_faithfulness: float | None
    ragas_answer_relevance: float | None
    latency_ms: float
    first_token_latency_ms: float
    input_tokens: int
    output_tokens: int
    estimated_cost: float
    refused: bool
    error: str | None


class ConfigMetricsResponse(BaseModel):
    name: str
    mode: SearchMode
    top_k: int
    rerank: bool
    case_count: int
    recall_at_5: float
    mrr: float
    faithfulness: float
    answer_relevance: float
    ragas_faithfulness: float | None
    ragas_answer_relevance: float | None
    average_latency_ms: float
    average_first_token_latency_ms: float
    average_input_tokens: float
    average_output_tokens: float
    average_estimated_cost: float
    refusal_rate: float
    hallucination_risk: float
    cases: list[CaseMetricsResponse]


class EvaluationRunAcceptedResponse(BaseModel):
    run_id: UUID
    dataset_name: str
    task_id: str
    status: EvaluationRunStatus
    attempt: int
    dataset_revision: int | None
    progress_completed: int
    progress_total: int


class EvalRunResponse(BaseModel):
    run_id: UUID
    dataset_name: str
    task_id: str | None
    status: EvaluationRunStatus
    attempt: int
    dataset_revision: int | None
    knowledge_base_revision: int
    progress_completed: int
    progress_total: int
    error_message: str | None
    created_at: str
    started_at: str | None
    completed_at: str | None
    results: list[ConfigMetricsResponse]


class EvaluationRunSummary(BaseModel):
    run_id: UUID
    dataset_name: str
    status: EvaluationRunStatus
    attempt: int
    dataset_revision: int | None
    knowledge_base_revision: int
    progress_completed: int
    progress_total: int
    created_at: str
